#! /bin/sh -e

# DP: ARM patch for const_double, courtesy Richard Earnshaw

src=gcc
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    src=$3/gcc
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p1 --fuzz 10 < $0
        cd $src && autoconf
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p1 --fuzz 10 < $0
        cd $src && autoconf
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0

--- src/gcc/config/arm/arm.c.old	Sun Jun 17 18:36:06 2001
+++ src/gcc/config/arm/arm.c	Sun Jun 17 18:36:40 2001
@@ -2416,6 +2416,11 @@
   if (GET_CODE (x) == SYMBOL_REF)
     return 1;
 
+  /* CONST_DOUBLE can contain a symbol ref for its in-memory
+     representation.  We are not interested about that case here.  */
+  if (GET_CODE (x) == CONST_DOUBLE)
+    return 0;
+
   fmt = GET_RTX_FORMAT (GET_CODE (x));
   for (i = GET_RTX_LENGTH (GET_CODE (x)) - 1; i >= 0; i--)
     {
